sensor/common/compliance/nodescan_handler_impl.go

```diff
@@ -43,40 +59,44 @@ func (c *nodeScanHandlerImpl) ProcessMessage(_ *central.MsgToSensor) error {
 	return nil
 }
 
-func (c *nodeScanHandlerImpl) run() {
-	defer c.stoppedC.SignalWithError(c.stopC.Err())
-	defer close(c.toCentral)
-
-	for !c.stopC.IsDone() {
-		select {
-		case <-c.stopC.Done():
-			return
-		case scan, ok := <-c.nodeScans:
-			if !ok {
-				c.stopC.SignalWithError(errors.New("channel receiving node scans v2 is closed"))
+// run handles the messages from Compliance and forwards them to Central
+// This is the only goroutine that writes into the toCentral channel, thus it is responsible for creating and closing that chan
+func (c *nodeScanHandlerImpl) run() <-chan *central.MsgFromSensor {
+	toC := make(chan *central.MsgFromSensor)
+	go func() {
+		defer c.runFinishedC.Signal()
+		defer close(toC)
+		for !c.stopC.IsDone() {
+			select {
+			case <-c.stopC.Done():
 				return
+			case scan, ok := <-c.nodeScans:
+				if !ok {
+					c.stopC.SignalWithError(errors.New("channel receiving node scans v2 is closed"))
+					return
+				}
+				// TODO(ROX-12943): Do something with the scan, e.g., attach NodeID
+				c.sendScan(toC, scan)
 			}
-			// TODO(ROX-12943): Do something with the scan, e.g., attach NodeID
-			c.sendScan(scan)
 		}
-	}
+	}()
+	return toC
 }
 
-func (c *nodeScanHandlerImpl) sendScan(scan *storage.NodeScanV2) {
-	if scan != nil {
-		select {
-		case <-c.stopC.Done():
-			return
-		case c.toCentral <- &central.MsgFromSensor{
-			Msg: &central.MsgFromSensor_Event{
-				Event: &central.SensorEvent{
-					Resource: &central.SensorEvent_NodeScanV2{
-						NodeScanV2: scan,
-					},
+func (c *nodeScanHandlerImpl) sendScan(toC chan *central.MsgFromSensor, scan *storage.NodeScanV2) {
+	if scan == nil {
+		return
+	}
+	select {
+	case <-c.stopC.Done():
```

Comment:
If I understand this correctly, there are two reasons why `c.stopC.Done()` would have stopped:

1) The parent channel (`c.nodeScans`) is closed
2) `Stop()` was called on the component. 

I'm not sure what is the meaning of calling `Stop()` on the component -- but I assume this would be a "graceful shutdown" of sensor. If that's the case, it's likely that other components (such as the one responsible for closing `c.nodeScans`) will also shutdown and close their channels. 

With that in mind, the current code will drop the events that are already enqueued and they will be lost. I'm not sure if this is a problem or not, but if these events can't be lost, it would be better to read the channel until it returns `nil, false`, so we know that all events are sent.

Reply:
I also spotted loss of events in some previous iteration on this code. I think what you suggest is correct, although achieving that will require changing the service part that receives messages from clients.

When the graceful shutdown is requested, the service should stop receiving (or ACK-ing) new messages from clients and close its internal channel (that's used for passing these messages further in Sensor's code). That should cascade throughout the system until all internal channels are closed so that the Central Sender also gets to understand that there will be nothing more to send and shuts down itself.

I'm afraid that such a change may turn quite a lot bigger than this PR. I think we should do it, but my vote for a separate PR. In this PR I suggest to not attempt fixing everything (especially deeper patterns such as `Start()` after `Stop()`), but rather just enough to make it not worse than before.

The specific message with node scan V2 (node inventory) can be dropped on the floor. By design, Compliance will resend it after 4 hours. This isn't great but it will be improved as part of the same epic by making Compliance retry sending until it receives an ACK - https://issues.redhat.com/browse/ROX-13164